### PR TITLE
refactor: refactor storage of cluster meta data

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Refer to #94 

# Rationale for this change
 Refactor storage of cluster meta data, define the memory data structure corresponding to `pb`. 
 

# What changes are included in this PR?
* Remove `Storage` and rename `MetaStorage` to `Storage`
* Refactor `MetaStorage` interface by using `Request` and `Result` to avoid multi parameters
* Define the memory data structure corresponding to `pb`
* Modify some log info to be more readable

# Are there any user-facing changes?
No.

# How does this change test
Old tests are covered.

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -267,33 +275,41 @@ func (c *Cluster) updateShardVersionLocked(shardID uint32, prevVersion, newVersi
 	}, nil
 }
 
-func (c *Cluster) getShardTopologyFromStorage(ctx context.Context, shardID uint32) (*clusterpb.ShardTopology, error) {
-	shardTopologies, err := c.storage.ListShardTopologies(ctx, c.clusterID, []uint32{shardID})
+func (c *Cluster) getShardViewFromStorage(ctx context.Context, shardID uint32) (*clusterpb.ShardTopology, error) {
```
Comment (ZuLiangWang) [inline server/cluster/cluster.go:278]: Same as above, there is two struct `ShardTopology` and `ShardView`, and they are often confused.
```suggestion
func (c *Cluster) getShardTopologyFromStorage(ctx context.Context, shardID uint32) (*clusterpb.ShardTopology, error) {
```

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -267,33 +275,41 @@ func (c *Cluster) updateShardVersionLocked(shardID uint32, prevVersion, newVersi
...
 	}
-	if len(shardTopologies) != 1 {
-		return nil, ErrGetShardTopology.WithCausef("shard has more than one shard topology, shardID:%d, shardTopologies:%v",
-			shardID, shardTopologies)
+	if len(shardViewsResult.ShardViews) != 1 {
+		return nil, ErrGetShardView.WithCausef("shard has more than one shard view, shardID:%d, shardViewsResult:%v",
+			shardID, shardViewsResult)
 	}
-	return shardTopologies[0], nil
+	shardView := storage.ConvertShardViewToPB(shardViewsResult.ShardViews[0])
```
Comment (ZuLiangWang) [inline server/cluster/cluster.go:290]:
```suggestion
	shardTopology := storage.ConvertShardViewToPB(shardViewsResult.ShardViews[0])
```

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -304,44 +320,48 @@ func (c *Cluster) createTableOnShardLocked(ctx context.Context, shardID uint32,
 	}, nil
 }
 
-func (c *Cluster) createTableInCache(shardID uint32, schema *Schema, tablePb *clusterpb.Table, prevVersion, newVersion uint64) (*Table, *ShardVersionUpdate, error) {
-	table := c.updateTableCacheLocked(shardID, schema, tablePb)
+func (c *Cluster) createTableInCache(shardID uint32, schema *Schema, tablePB *clusterpb.Table, prevVersion, newVersion uint64) (*Table, *ShardVersionUpdate, error) {
```
Comment (ZuLiangWang) [inline server/cluster/cluster.go:323]: Consider encapsulation all logic about cache as struct `ClusterCache` ?

File: server/storage/types.go [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+type GetTableRequest struct {
+	ClusterID ClusterID
+	SchemaID  SchemaID
+	TableName string
+}
+
+type GetTableResult struct {
+	Table  Table
+	Exists bool
```
Comment (ZuLiangWang) [inline server/storage/types.go:87]: Why add this variable `Exists`? maybe we can just check table is exists like this:
```
GetTableResult.Table == nil
```

File: server/storage/storage_impl.go [hunk 5]
```diff
@@ -40,41 +38,41 @@ func newEtcdStorage(client *clientv3.Client, rootPath string, opts Options) Stor
...
 }
 
-func (s *metaStorageImpl) ListClusters(ctx context.Context) ([]*clusterpb.Cluster, error) {
+func (s *metaStorageImpl) ListClusters(ctx context.Context) (ListClustersResult, error) {
 	startKey := makeClusterKey(s.rootPath, 0)
 	endKey := makeClusterKey(s.rootPath, math.MaxUint32)
 	rangeLimit := s.opts.MaxScanLimit
 
-	clusters := make([]*clusterpb.Cluster, 0)
+	clusters := make([]Cluster, 0)
```
Comment (ZuLiangWang) [inline server/storage/storage_impl.go:46]:
```suggestion
	clusters := []Cluster{}
```

File: server/storage/storage_impl.go [hunk 6]
```diff
@@ -40,41 +38,41 @@ func newEtcdStorage(client *clientv3.Client, rootPath string, opts Options) Stor
...
-	cluster.CreatedAt = uint64(now.Unix())
-
-	value, err := proto.Marshal(cluster)
+// CreateCluster return error if the cluster already exists.
+func (s *metaStorageImpl) CreateCluster(ctx context.Context, req CreateClusterRequest) error {
+	c := convertClusterToPB(req.Cluster)
+	value, err := proto.Marshal(&c)
 	if err != nil {
-		return nil, ErrDecodeCluster.WithCausef("fail to decode cluster，clusterID:%d, err:%v", cluster.Id, err)
+		return ErrDecodeCluster.WithCausef("encode cluster，clusterID:%d, err:%v", req.Cluster.ID, err)
```
Comment (ZuLiangWang) [inline server/storage/storage_impl.go:72]:
```suggestion
		return ErrEncodeCluster.WithCausef("encode cluster，clusterID:%d, err:%v", req.Cluster.ID, err)
```

Comment (ZuLiangWang) [reply server/storage/storage_impl.go:72]: Check `ErrEncodeCluster` and `ErrDecodeCluster` in all files.

File: server/storage/storage_impl.go [hunk 7]
```diff
@@ -85,129 +83,129 @@ func (s *metaStorageImpl) CreateCluster(ctx context.Context, cluster *clusterpb.
...
-	clusterTopology.CreatedAt = uint64(now.Unix())
-
-	value, err := proto.Marshal(clusterTopology)
+// CreateClusterView return error if the cluster view already exists.
+func (s *metaStorageImpl) CreateClusterView(ctx context.Context, req CreateClusterViewRequest) error {
+	pb := ConvertClusterViewToPB(req.ClusterView)
+	value, err := proto.Marshal(&pb)
 	if err != nil {
-		return nil, ErrDecodeClusterTopology.WithCausef("fail to decode cluster topology, clusterID:%d, err:%v", clusterTopology.ClusterId, err)
+		return ErrDecodeClusterView.WithCausef("encode cluster view, clusterID:%d, err:%v", pb.ClusterId, err)
```
Comment (ZuLiangWang) [inline server/storage/storage_impl.go:99]: Same as above.

File: server/storage/storage_impl.go [hunk 8]
```diff
@@ -85,129 +83,129 @@ func (s *metaStorageImpl) CreateCluster(ctx context.Context, cluster *clusterpb.
...
 
-// Return error if the cluster topology already exists.
-func (s *metaStorageImpl) CreateClusterTopology(ctx context.Context, clusterTopology *clusterpb.ClusterTopology) (*clusterpb.ClusterTopology, error) {
-	now := time.Now()
-	clusterTopology.CreatedAt = uint64(now.Unix())
-
-	value, err := proto.Marshal(clusterTopology)
+// CreateClusterView return error if the cluster view already exists.
+func (s *metaStorageImpl) CreateClusterView(ctx context.Context, req CreateClusterViewRequest) error {
+	pb := ConvertClusterViewToPB(req.ClusterView)
```
Comment (ZuLiangWang) [inline server/storage/storage_impl.go:96]: I think this variable name is fuzzy, maybe rename it as `ClusterTopology` will better.

File: server/cluster/manager.go [hunk 9]
```diff
@@ -331,14 +343,14 @@ func (m *managerImpl) Start(ctx context.Context) error {
...
 		if err := cluster.Load(ctx); err != nil {
-			log.Error("cluster manager fail to start, fail to load cluster", zap.Error(err))
-			return errors.WithMessagef(err, "cluster manager start, clusters:%v", cluster)
+			log.Error("fail to load cluster", zap.String("cluster", cluster.Name()), zap.Error(err))
+			return errors.WithMessagef(err, "fail to load cluster:%v", cluster.Name())
 		}
+		log.Info("open cluster successfully", zap.String("cluster", cluster.Name()))
 		m.clusters[cluster.Name()] = cluster
 	}
 	m.running = true
```
[the remark this answers is not in the document]
Comment (chunshao90) [reply server/cluster/manager.go:356]: The last setting of true here is to ensure that the manager is running after all clusters are opened.

File: server/coordinator/procedure/scatter.go [hunk 10]
```diff
@@ -63,16 +63,16 @@ func scatterPrepareCallback(event *fsm.Event) {
 		return
 	}
 
-	shardTopologies := make([]*clusterpb.ShardTopology, 0, len(shards))
+	shardViews := make([]*clusterpb.ShardTopology, 0, len(shards))
```
[the remark this answers is not in the document]
Comment (chunshao90) [reply server/coordinator/procedure/scatter.go:66]: I will rename `ShardTopology` to `ShardView` in `pb` later.

File: server/cluster/cluster.go [hunk 11]
```diff
@@ -18,12 +19,12 @@ import (
 )
 
 type metaData struct {
-	cluster         *clusterpb.Cluster
-	clusterTopology *clusterpb.ClusterTopology
+	cluster     storage.Cluster
```
[the remark this answers is not in the document]
Comment (chunshao90) [reply server/cluster/cluster.go:22]: Cluster info will not be shared and modified, so no pointer is needed for the time being.

File: server/storage/types.go (hunk 4 above)
Comment (chunshao90) [reply server/storage/types.go:87]: `GetTableResult.Table` is a struct, can't use `GetTableResult.Table == nil`.

File: server/storage/storage_impl.go (hunk 8 above)
Comment (chunshao90) [reply server/storage/storage_impl.go:96]: I rename it to `clusterViewPB`.

**Review by ZuLiangWang — APPROVED** [review]
LGTM

**Pull request merged**